pkg/scripts/install.go

```diff
@@ -0,0 +1,173 @@
+// map of scripts with detailed information about each of the scripts install process
+// with error if installing some of them failed
+func (i *installer) InstallRequired(scripts []api.Script, dstDir string) (results map[api.Script]*InstallResult, err error) {
+	results = i.run(scripts, dstDir)
+	failedScripts := []api.Script{}
+	for script, result := range results {
+		if !result.Installed && result.Error != nil {
+			failedScripts = append(failedScripts, script)
+		}
+	}
+	if len(failedScripts) > 0 {
+		err = errors.NewInstallRequiredError(failedScripts)
+	}
+
+	return
+}
+
+// InstallOptional downloads and installs a set of scripts into dstDir, the result is a
+// map of scripts with detailed information about each of the scripts install process
+func (i *installer) InstallOptional(scripts []api.Script, dstDir string) map[api.Script]*InstallResult {
+	return i.run(scripts, dstDir)
+}
+
+func (i *installer) run(scripts []api.Script, dstDir string) map[api.Script]*InstallResult {
+	results := make(map[api.Script]*InstallResult)
+	if len(scripts) == 0 {
+		return results
+	}
+	for _, script := range scripts {
+		results[script] = &InstallResult{}
+	}
+
+	//get scripts from parameter URL
+	if i.scriptsURL != "" {
+		i.download(i.scriptsURL, results, filepath.Join(dstDir, string(api.DownloadedScripts)))
+	}
+
+	// get scripts from default URL
+	defaultURL, err := i.docker.GetScriptsURL(i.image)
+	if err == nil && defaultURL != "" {
+		i.download(defaultURL, results, filepath.Join(dstDir, string(api.DefaultScripts)))
+	} else {
+		glog.Warningf("Image %s does not contain STI_SCRIPTS_URL defined!", i.image)
+	}
+
+	i.install(results, dstDir)
+	return results
+}
+
+func (i *installer) download(scriptsURL string, scripts map[api.Script]*InstallResult, dstDir string) {
+	for script, result := range scripts {
+		url, err := url.Parse(scriptsURL + "/" + string(script))
+		if err != nil {
+			result.Error = err
+			continue
+		}
+
+		target := filepath.Join(dstDir, string(script))
+		download, err := i.downloader.Download(url, target)
+		// make sure to have appropriate file permissions for downloaded scripts
+		if download && err == nil {
+			if err = i.fs.Chmod(target, 0700); err != nil {
+				result.Error = err
+			}
+		}
+		// save information about download & error only if there's no  previous information
+		if !result.Downloaded {
+			result.Downloaded = download
+		}
+		if result.Error == nil {
+			result.Error = err
+		}
+	}
+}
+
+func (i *installer) install(results map[api.Script]*InstallResult, dstDir string) {
+	for script, result := range results {
+		// ignore not downloaded scripts
+		if !result.Downloaded && result.Error == nil {
+			result.Installed = true
+			continue
+		}
+
+		scriptPath := i.getScriptPath(script, dstDir)
+		if scriptPath == "" {
+			result.Error = errors.NewInstallError(script)
+			continue
+		}
+		if err := i.fs.Rename(scriptPath, filepath.Join(dstDir, "upload/scripts", string(script))); err != nil {
```

Comment:
This is the actual install part, we're picking a script and we're moving it to `upload` directory. Without it you won't get any scripts uploaded to builder image.